### ISSUE TITLE
useQuery randomly not returning data and stuck in loading

### ISSUE THREAD

**Issue by vsaarinen** [tannerlinsley/react-query#1657]
**Describe the bug**
After upgrading to react-query 3.5.11 from 2.26.4 last week, our project's Cypress tests started failing randomly because the app started showing a loading spinner at random times. After a lot of head-banging, we finally tracked down the issue: even though, based on the network inspector, the request had gone through successfully, react-query's `useQuery` was returning an object with `isLoading` as `true` and `data` as `undefined`. Reverting back to react-query 2 fixed the problems.

Most of the requests in our Cypress tests are stubbed using `cy.server` and `cy.route`. This probably shouldn't affect react-query but mentioning it here just in case.

Our cache is configured in the following manner at the top of one of our source files, so this shouldn't be caused by us accidentally creating multiple caches:

```typescript
const queryCache = new QueryCache({
  defaultConfig: {
    queries: {
      staleTime: 10 * 60 * 1000,
      cacheTime: 15 * 60 * 1000,
      retry: (
        _failureCount,
        error: any,
      ) => error.status !== 403 && error.status !== 401,
    },
  },
});

// Further down:

class AppProvider extends Component<{}, State> {
  public render() {
    return (
      <ReactQueryCacheProvider queryCache={queryCache}>
        {/* ... */}
      </ReactQueryCacheProvider>
    );
  }
}
```

**To Reproduce**
Unfortunately, this happens randomly and is thus very hard to reproduce. The randomness does point towards some type of race condition. Our app performs many (~10) requests on startup.

Even though which tests fail is random, I can consistently get at least one test to fail. Is there something I can do to help track down what could be the issue?

**Expected behavior**
`useQuery` should return the data it receives in the response once the request is successful.

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Chrome
 - Version: 87

**Additional context**
Happens on both Cypress version 5.3.0 and 6.2.1.

**Comment by TkDodo** [comment]
this looks surprisingly similar to https://github.com/tannerlinsley/react-query/issues/1657

**Comment by boschni** [comment]
Don't know how the tests are written but it might be good to check if they do not implicitly depend on certain timings. Are you able to reproduce it manually? This might be useful: https://docs.cypress.io/guides/core-concepts/retry-ability.html

**Comment by vsaarinen** [comment]
> Don't know how the tests are written but it might be good to check if they do not implicitly depend on certain timings. Are you able to reproduce it manually? This might be useful: https://docs.cypress.io/guides/core-concepts/retry-ability.html

I haven't noticed this problem when using the app manually, so there's always a chance that this is somehow caused by Cypress. But a few reasons why I think this may be an issue with react-query:

- The error (loading spinner) occurs on app startup, before any interaction from Cypress. So it shouldn't be due to timing of commands/assertions.
- The problem goes away when react-query is downgraded to an old version.
- When inspecting the app, the network requests show that the response was received by the app but the related `useQuery` call returns `undefined` for `data`.

I'll see if I can narrow down the react-query version where this started happening.

**Comment by antonversal** [comment]
I can confirm the same behavior after migrating from 2.23.1 to 3.5.16.
`isLoading` is randomly returning true for queries with `staleTime` option. Everything works without `staleTime`.

**Comment by boschni** [comment]
Would be great if you guys could to provide a sandbox or something similar which we can use to debug the issue.

**Comment by antonversal** [comment]
> Would be great if you guys could to provide a sandbox or something similar which we can use to debug the issue.

I've tried to reproduce in a sandbox without any luck (

**Comment by josewhitetower** [comment]
Commenting just to subscribe to the thread because I'm having the same issue and can't reproduce it on a sandbox. Happens  on v2.26.3 and after updating to v3.6.0.

**Comment by TkDodo** [comment]
Can we try to narrow it down collectively? So far, we've seen that:

- it apparently still works on `2.23.1` according to @antonversal 
- it's no longer working on `2.26.3` according to @josewhitetower 

could you further narrow it down to a specific version between those two?

`staleTime` seems to be involved, but how do you set staleTime?  via the default-config on the global cache like @vsaarinen has shown in the opening post (similar to what @fTrestour has posted in #1653), or are you setting the `staleTime` on the `useQuery` call directly?

**Comment by TkDodo** [comment]
I'm also seeing that both the opening post in this issue as well as in #1653 have the `retry` option set - could we rule that one out as well?

**Comment by antonversal** [comment]
@TkDodo I set `staleTime` in `useQuery`. No global settings and `retry`.

**Comment by vsaarinen** [comment]
> Can we try to narrow it down collectively? So far, we've seen that:
> 
> * it apparently still works on `2.23.1` according to @antonversal
> * it's no longer working on `2.26.3` according to @josewhitetower
> 
> could you further narrow it down to a specific version between those two?

For me, things work fine on `2.26.4`. I'll try to find which version of 3 causes things to start breaking again.

**Comment by vsaarinen** [comment]
Started seeing the same issues after upgrading to `3.2.0`. So it appears to be a regression introduced in version 3.

**Comment by adeelibr** [comment]
> Started seeing the same issues after upgrading to `3.2.0` from `2.26.4`. So it appears to be a regression introduced in version 3.

I have the same issue

**Comment by fermmm** [comment]
I have the same issue, at leat in my case I found this: 
When is stuck it seems it's because the refectch is returning the same object, when the query is stuck this console.log prints true on the console:
const { data, refetch } = useQuery(...)
... 
refetch().then(response => console.log(response.data === data));

**Comment by fermmm** [comment]
> I have the same issue, it's doing it randomly like the others are reporting. I've found this:
> 
> When it's stuck refetch() returns the same object, when the query is stuck this console.log prints true and when it's not stuck prints false:
> 
> ```
> const { data, refetch } = useQuery(...)
> ... 
> refetch().then(response => console.log(response.data === data));
> ```
> 
> Calling refetch() again when the data is the same doesn't work, returns the same data again, so there is no workaround possible yet.
> 
> **Edit: Upgraded from 3.5.1 to 3.6.0 and the problem is still happening but at least this time calling refetch() again when the query is stuck doesn't return the same object so it's a workaround:**
> 
> ```
> const { data, refetch } = useQuery(...)
> ...
> refetch().then(response => {
>     if(response.data === data){
>         refetch()
>     }
> });
> ```
> 
> I have staleTime: Infinity on the new QueryClient() (global setting)
> Using version: 3.5.1
> Reproducing this on a sandbox takes a lot of effort and time I don't have. This bug already took a lot of my time and it's a blocker for me.

Self-response: I finally found the cause at least for my specific case.

React-query performs a deep compare to update the component, so if your request returns the same data than in the previous request, react query will not update your component. This was happening randomly because the response of my request contains random changes and sometimes returns the same data. So if you trigger a request you should not expect a component re-render unless the returned data from the server is different according to the deep compare function of react-query. In my case the returned data was not different and I was expecting a re-render because I have my own logic to deal with the server returning the same data.

The solution in case you have the same problem is to disable deep compare setting structuralSharing to false like this:

```
export const queryClient = new QueryClient({
   defaultOptions: {
      queries: {
         staleTime: Infinity,
         structuralSharing: false
      },
      mutations: {}
   }
});
```

This feature was introduced because of this issue:
https://github.com/tannerlinsley/react-query/issues/430

**Comment by vsaarinen** [comment]
@fermmm It sounds like your issue may be different than what I'm experiencing. My issue is that even though data was successfully received by the app, `useQuery` thinks that it's still loading the data and never provides it to the app.

**Comment by fermmm** [comment]
> @fermmm It sounds like your issue may be different than what I'm experiencing. My issue is that even though data was successfully received by the app, `useQuery` thinks that it's still loading the data and never provides it to the app.

Yes my problem looks similar but may be not the same. Unless the response of your server is empty and there is no cache, so de deep compare function returns that the results are the same, that could be triggering your specific bug. I bet is also caused by deep compare in some way. Try disabling structuralSharing.

**Comment by fermmm** [comment]
Ok I had this issue in another request of my app and this time was not related with structuralSharing, I noticed {data: undefined,  isLoading: true} randomly without getting the data. The only different thing with this request it was that I had many components mounted that use this query, maybe is something that happens when one component is waiting for the request and other mounts with the same query... I don't know.

I decided to port all my requests to SWR, I tested it a lot and looks really stable, they don't have issues in their github like this one, where random instability happens and nobody can reproduce. Also I noticed they implemented some features first, like the deep equality, also I trust the next.js devs.

**Comment by vasconita** [comment]
Same problem here. In my case I use `staleTime: Infinity`.

I tested @fermmm workaround and it seems to work, thanks!

**Comment by TkDodo** [comment]
I would really like to get to the bottom of this issue, but if nobody can reproduce it in a codesandbox, it is almost impossible to narrow down.

**Comment by vasconita** [comment]
I forgot to add more info: I'm mounting a lot of components (50) that use this `useQuery` with `Infinity` `staleTime` and I don't have configured `retry` neither. I'm going to try to reproduce it in a sandbox.

**Comment by vasconita** [comment]
I have created a [sandbox](https://codesandbox.io/s/staletimeinfinite-i9lne?file=/src/index.js) with the conditions that seem to be common to all of us (Infinity staleTime and having many instances) but cannot reproduce the error, I think it has only happened to me once. On my SPA it happens to me much more easily 🤔

**Comment by TkDodo** [comment]
From what I can see in the sandbox: are you also putting the whole result from `fetch` into the query cache in your app? This includes some non-serializable data, like the `body`, which is a stream, and lots of functions that return Promises, like `.json()`...

**Comment by vasconita** [comment]
No no, it was like this because since I don't use `data` in the example I didn't even worry with the response. I updated it and now I just return the JSON body (like in my app of course)

**Comment by fermmm** [comment]
In my case this only happens when I restart my RN app when all the components are mounting for the first time

**Comment by vsaarinen** [comment]
I believe my issue has also appeared only on app start, in the initial loading of components.

**Comment by TkDodo** [comment]
> I believe my issue has also appeared only on app start, in the initial loading of components.

I guess this is the only time when you can be "stuck in loading", because you are never in loading state again once you've successfully loaded data.

Question: In case this being-stuck happens, what do the react-query devtools say about the state of the query?

**Comment by vsaarinen** [comment]
> Question: In case this being-stuck happens, what do the react-query devtools say about the state of the query?

The problematic query is the second from the bottom in the list, with 5 listeners and marked as stale:
<img width="1186" alt="Screenshot 2021-02-02 at 8 54 37" src="https://user-images.githubusercontent.com/888333/106563516-8b050980-6534-11eb-8744-aa1deec8a764.png">
<img width="1187" alt="Screenshot 2021-02-02 at 8 54 47" src="https://user-images.githubusercontent.com/888333/106563542-97896200-6534-11eb-929c-f71d45cced4f.png">

I confirmed that the data for that query was received successfully when looking at the network developer tab. The weird thing is that the data is shown correctly in the devtools, but in our app the `data` property of the objected returned by that `useQuery` is undefined.

**Comment by TkDodo** [comment]
> The weird thing is that the data is shown correctly in the devtools, but in our app the data property of the objected returned by that useQuery is undefined.

Yes, that does indeed seem weird. Are you using different `staleTimes` on the  5 observers? Everyone is mentioning `staleTime: Infinity`, but in your code, I am seeing a different `staleTime` - and your query is also marked as `stale`, which shouldn't be the case with `staleTime: Infinity`...

**Comment by vsaarinen** [comment]
> Yes, that does indeed seem weird. Are you using different `staleTimes` on the 5 observers? Everyone is mentioning `staleTime: Infinity`, but in your code, I am seeing a different `staleTime` - and your query is also marked as `stale`, which shouldn't be the case with `staleTime: Infinity`...

I've never used `staleTime: Infinity`. Except for a couple of queries where these are overridden, I only use the time values from the `defaultConfig` as shown in the original post:
```
      staleTime: 10 * 60 * 1000,
      cacheTime: 15 * 60 * 1000,
```

**Comment by TkDodo** [comment]
> I've never used staleTime: Infinity.

thanks, I mixed that up a bit 😅 .

>  since I don't know what "stale" means.

a `stale` query is a query that will refetched in the background if a new subscriber mounts / the window is refocussed etc.

>  the devtools marks the query as stale

judging from the code, the devtools don't manipulate the queries (unless you press one of the buttons). They just display the current state of the cache.

---

So if  your screenshot is from after the 5 subscribers mount, and you have a `staleTime` of 10 minutes - why are the queries showing up as `stale` ? They should show up as `fresh` (green) - just like the other queries 🤔

**Comment by vsaarinen** [comment]
> So if your screenshot is from after the 5 subscribers mount, and you have a `staleTime` of 10 minutes - why are the queries showing up as `stale` ? They should show up as `fresh` (green) - just like the other queries 🤔

Good question. The top 6 queries in the list _should_ be marked as fresh. Maybe this is related to the underlying cause?

**Comment by TkDodo** [comment]
> Good question. The top 6 queries in the list should be marked as fresh. Maybe this is related to the underlying cause?

yes, maybe. Are you sure that _all_ subscribers (all instances of `useQuery`) use _the same_ config? If you have 4 subscribers with staleTime: 10 minutes, and one additional subscriber where the staleTime is 0, the smaller one will "win". Not sure if that could cause this problem ...

**Comment by vsaarinen** [comment]
> Are you sure that _all_ subscribers (all instances of `useQuery`) use _the same_ config? If you have 4 subscribers with staleTime: 10 minutes, and one additional subscriber where the staleTime is 0, the smaller one will "win". Not sure if that could cause this problem ...

All of the subscribers listed in the screenshots use the default `cacheTime` of 15 minutes and `staleTime` of 10 minutes.

**Comment by boschni** [comment]
Seems like the cache is fine, but maybe one of the observers did not get updated somehow. Can you see in the devtools if one of the observers for that query has an incorrect `currentResult`?

**Comment by vsaarinen** [comment]
> Seems like the cache is fine, but maybe one of the observers did not get updated somehow. Can you see in the devtools if one of the observers for that query has an incorrect `currentResult`?

Yep, there's an issue with the observers: the first three have the correct data in `currentResult` but the last one has invalid data, as shown in the screenshots below.

<img width="983" alt="Screenshot 2021-02-02 at 12 01 41" src="https://user-images.githubusercontent.com/888333/106584213-9bc27900-654e-11eb-8627-d65b1f2b1136.png">

<img width="984" alt="Screenshot 2021-02-02 at 12 01 54" src="https://user-images.githubusercontent.com/888333/106584222-9ebd6980-654e-11eb-815e-2c766ef7b5f9.png">

**Comment by boschni** [comment]
Just guessing here, but it seems technically possible that a query gets updated between a component mount and its useEffect.

1. Component mount
2. Query update
3. Component useEffect (starts listening for updates)

In such case the component would miss the update.

Could you check if this change fixes it?

```diff
--- a/src/core/queryObserver.ts
+++ b/src/core/queryObserver.ts
@@ -528,6 +528,11 @@ export class QueryObserver<
       )
 
     if (query === prevQuery) {
+      if (query.state !== this.currentResultState) {
+        const prevResult = this.currentResult
+        this.updateResult()
+        if (prevResult !== this.currentResult) {
+          this.notify({ listeners: true })
+        }
+      }
       return
     }
```

**Comment by vsaarinen** [comment]
> Could you check if this change fixes it?

After a few quick tests, it does seem like this fixes the issue. 🎉 But need to do a bit more testing...

I added the following console logging to make sure that the condition gets hit, and the weird thing is that the `this.notify()` never seems to be get called but the issue still seems to be fixed.

```typescript
    if (query === prevQuery) {
      if (query.state !== this.currentResultState) {
        console.warn('!!! query.state !== this.currentResultState !!!')
        const prevResult = this.currentResult
        this.updateResult()
        if (prevResult !== this.currentResult) {
          console.warn('!!! prevResult !== this.currentResult !!!')
          this.notify({ listeners: true })
        }
      }
      return
    }
```

**Comment by maybesmurf** [comment]
> I can confirm the same behavior after migrating from 2.23.1 to 3.5.16.
> `isLoading` is randomly returning true for queries with `staleTime` option. Everything works without `staleTime`.

I'm currently experiencing this same issue. I opened a discussion on this yesterday. [Here](https://github.com/tannerlinsley/react-query/discussions/1732)

**Comment by boschni** [comment]
Could you check if the issue is fixed in v3.6.1?

**Comment by vsaarinen** [comment]
Haven't run into failing tests after upgrading to 3.6.1. Let's keep our fingers crossed.

Thanks a lot for the fix!

**Comment by vasconita** [comment]
It seems to solve the problem in my case. Thank you very much!!

**Comment by antonversal** [comment]
Thanks, @boschni. I do not see failing tests anymore as well.

**Comment by ivillarreal91** [comment]
Hello everyone!!!

Has anyone experienced the same behavior but now with `useQueries hook`? 
it doesn't return data and gets stuck in loading randomly in the tests. I appreciate any support 

```
"react": "^16.13.1"
"react-query": "^3.13.4"
```

**Comment by cameronevans** [comment]
@ivillarreal91 I'm seeing this as well with latest, were you able to find a solution? Using `useQuery` in one component and `useQueries` with the same key in another component causes the `useQuery` to get stuck. Going to try to reproduce in a CSB.

**Comment by Toyurc** [comment]
> @ivillarreal91 I'm seeing this as well with latest, were you able to find a solution? Using `useQuery` in one component and `useQueries` with the same key in another component causes the `useQuery` to get stuck. Going to try to reproduce in a CSB.

I think this is the closest thing that describes the problem, I am currently having a similar issue and I just upgraded to the latest release. if you have keys like `baa` and `baaFoo`, the whole pages hangs and is stuck on isFetching

**Comment by TkDodo** [comment]
@Toyurc if you have a codesandbox reproduction, please open a new issue with it.

**Comment by MatkoMilic** [comment]
I have the same issue, its stuck at loading, while the exact same code works on another app.

**Comment by TkDodo** [comment]
@MatkoMilic again, this is not very helpful I'm afraid. Without an isolated reproduction, there is only so much I can do 🤷

**Comment by ZiedHf** [comment]
@TkDodo I am also facing the same issue and stuggling to reproduce it.

**Comment by nawfalhaddi** [comment]
I also have the same issue in my react native application, it is a random issue, sometimes isLoading and isFetching get stuck and data stays undefined even if the network request is done successfully.

**Comment by nawfalhaddi** [comment]
> @MatkoMilic again, this is not very helpful I'm afraid. Without an isolated reproduction, there is only so much I can do 🤷

Hello, I'm running 4 API calls in the same component with different keys using "useQuery"  and I used QueriesObserver to see what is happening in the background. 
Basically, in the observer, the data is up to date and the isLoading state has changed to false, but this does not reflect on the variables which are extracted from useQuery.
is there a way to force a useQuery to check the observer again?

@TkDodo

**Comment by TkDodo** [comment]
@nawfalhaddi if its related to 4 useQuery queries running in parallel, can you reproduce this in a sandbox?

> Btw this is on a react native project

Are all problems posted in this issue related to react native?

**Comment by nawfalhaddi** [comment]
@TkDodo  I tried to reproduce the issue on sandbox but it's not happening, maybe it's related to react native specifically. 
I will create an example repo of react native project and try to reproduce the problem on it, I will share it with you soon.

**Comment by FeliceGeracitano** [comment]
could this be related to https://github.com/tannerlinsley/react-query/issues/3332 ?

**Comment by vdutz** [comment]
@TkDodo I am experiencing the same issue on version 3.24.4 - `isLoading` gets stuck on true. I'm using React Native and this happens to me randomly (but only on app startup). It seems to happen about every 1 in 10 or 20 times, so is quite difficult to replicate.

Any update on this?

I also have the same question as @nawfalhaddi:

>  is there a way to force a useQuery to check the observer again?

**Comment by theblinkdate** [comment]
Hey y'allz. I've been dealing with this for 8 hours myself and this is my story: 

My query that's stuck in loading has an `enabled` property, and I had a race condition where:
1. Component would get `enabled` set to true, so query fires
2. I have an axios interceptor that calls a function that updates the `enabled` property above
3. Before the API call finishes its normal lifecycle (succeed or fail), `enabled` is set to false before `isLoading` is flipped back to false. Basically the query gets disabled in a "loading" state.

For me I just needed to call `resetQueries`, but the race condition made the problem intermittent and it would only happen on app startup. Hopefully it helps somebody

**Comment by gus-delpino** [comment]
Hey y'allz. I've been dealing with this for 8 hours myself and this is my story:

My query that's stuck in loading has an enabled property, and I had a race condition where:
1. Component would get enabled set to true, so query fires
2. I have an axios interceptor that calls a function that updates the enabled property above
3. Before the API call finishes its normal lifecycle (succeed or fail), enabled is set to false before isLoading is flipped back to false. Basically the query gets disabled in a "loading" state.

For me I just needed to call resetQueries, but the race condition made the problem intermittent and it would only happen on app startup. Hopefully it helps somebody

**Comment by SevenOutman** [comment]
I'm experiencing the same issue on version 3.38.0. It seems to only happen when my Jest tests are run in CI environment with multiple workers (with `--maxWorkers=4` flag). It doesn't happen on my local machine, nor when I set `--maxWorkers=1` (single worker mode). DN if the info helps.

**Comment by TkDodo** [comment]
imo it doesn't make much sense if more people keep commenting on this closed issue from over a year ago. If you have a consistent reproduction, or any reproduction for that matter, please open a new issue. There isn't much I can do otherwise I'm afraid.

**Comment by balteo** [comment]
Hi @TkDodo 

I have implemented a small application that reproduces the issue described in this discussion. Here is the repo: https://github.com/balteo/react-query-issue-reproduction

Just git clone the repos and `yarn install` and `yarn run test`

Please let me know if I can help further.

Kind regards

**Comment by TkDodo** [comment]
I renamed the port-to-v3 label as port-to-v4, that was probably not a good idea, sorry.

code is quite different now, I would need to see a codesandbox reproduction that fails on v4 please.

the code path you're showing has been removed in a later v3 version to ensure concurrent rendering capabilities:
- https://github.com/TanStack/query/pull/1775/files#diff-93f379800fc8abf895eba249b2e2371eda98740aa40fc9f284a8088d190f46c3L97

**Comment by annjawn** [comment]
I am going to add a +1 to this. I am facing this exact issue where `isLoading` gets stuck in `true` and `data` comes back as `undefined`.  I am using the same query in two components the first component fires the query and gets the data fine, but on the second component the same query (with the same key) gets stuck and returns undefined. This is a bizarre situation that I am not sure I am able to wrap my head around. 

Here's whats happening

```
Component - 1
    Query 1 --> Works 
    Query 2 --> Works

Component - 2
   Query 1 --> Works with same key as in component 1
   Query 2 --> Doesn't Work with same or different key
```

**Comment by annjawn** [comment]
ok sheesh i think i found the issue. With Multiple queries comes multiple `isLoading` the part of the component that depended on the second query was rendering too soon causing the issue. So these multiple `isLoadings` and checking of `data` for each of these queries will need to be handled individually. Any early rendering causes the query status to stay stuck (or rather give the illusion i should say).

**Comment by filip-kis-fsc** [comment]
I am experiencing this issue consistently in a larger project with 
```
    "@tanstack/react-query": "4.26.0",
    "react": "18.2.0",
    "next": "13.2.3",
```
but only if react is running in strict mode.

Unfortunately I do not have a minimal repro example.

**Comment by hectopod** [comment]
I am facing the same issue only in Cypress. 

Devtools indicate that the data is fresh but the loader is still loading (isFetching).

**Comment by xxRockOnxx** [comment]
@agos thanks for the clue about invalidating a query being the cause. in my case im doing `cancelQueries` instead. i tried adding `await` in front and now it works! not sure if this helps.

**Comment by crutchcorn** [comment]
One thing to remember if you're in the MSW/Testing land - set `retry` to `0` during testing.

Remember that Query intentionally backs off over a period of time to retry searching, upwards of a couple of seconds. **This is a good thing, as it leads to better UX**.

As a result, my tests were failing until I did this.

**Comment by vaibhavshn** [comment]
We also faced the same issue where for some reason some users were not able to fetch the data properly.

On further debugging found the fetchStatus was always `paused`. After going through docs, we passed the `networkMode: always` and it worked fine this time. We also verified this by removing the networkMode config.

So this probably means something wrong is happening in the code when the default `networkMode: on` value is set.

cc @TkDodo

**Comment by TkDodo** [comment]
> So this probably means something wrong is happening in the code 

no, this is an issue with `navigator.onLine` in chrome on macos, which we are using in v4 to determine if you are online or not.

- https://github.com/TanStack/query/issues/5644
- https://github.com/TanStack/query/issues/5679#issuecomment-1623038531

Since `navigator.onLine` is broken, we've moved away from using it in v5

**Comment by vaibhavshn** [comment]
> Since navigator.onLine is broken, we've moved away from using it in v5

Thanks for the quick reply, will upgrade to v5.

**Comment by Dougmarcolino** [comment]
I just needed to upload it to the new version of Chrome, now it's working.

![image](https://github.com/TanStack/query/assets/7004832/4c73b747-0d15-46fe-baa5-a47a42664190)

macOS: Ventura 13.5.2

**Comment by Harvel-Kay** [comment]
See am encountering the same issue with v4.36,
It tends to make requests successfully to jsonplaceholder however when I make a request to my local server it then stucks on the loading,
When I use the use Effect hook and state everything works well, I've seen this stuck issue has been perturbing from v2 all the way to v4, why on earth haven't you figured out why your library behaves like this, it's pretty frustrating and I'll chose to revert to the older way of fetching data 😡😡😡😡😠😠

**Comment by TkDodo** [comment]
> It's quite strange.

again, I'm positive that this is related to `navigator.onLine` being broken in chrome, which is not our fault. What is our fault is that we were relying on it to work properly in the first place, which is what we've changed in v5.

you can validate that assumption by looking at the `fetchStatus` returned from `useQuery` when that happens. If the `fetchStatus` is `'paused'`, you'll know that this is indeed the case.

again, you can:

- update to v5, which has a different default behaviour
- set `networkMode: 'offlineFirst'` to work around it
- get the bugs fixed in chrome 🤷

**Comment by hoanmq** [comment]
> > It's quite strange.
> 
> again, I'm positive that this is related to `navigator.onLine` being broken in chrome, which is not our fault. What is our fault is that we were relying on it to work properly in the first place, which is what we've changed in v5.
> 
> you can validate that assumption by looking at the `fetchStatus` returned from `useQuery` when that happens. If the `fetchStatus` is `'paused'`, you'll know that this is indeed the case.
> 
> again, you can:
> 
> * update to v5, which has a different default behaviour
> * set `networkMode: 'offlineFirst'` to work around it
> * get the bugs fixed in chrome 🤷

Thank you for responding, however, it happens randomly across different browsers including Edge, Brave, etc.
 I will try upgrading to v5.

**Comment by TkDodo** [comment]
edge and brave are both built on chromium ... they are the same thing and have the same behaviour / bugs in this regard

**Comment by 117** [comment]
Firefox 120.0.1 (64-bit) still occurring  M2 Mac, Sonoma 14.1.1

**Comment by TkDodo** [comment]
@117 on query v4 or v5? We haven't done anything on v4 - and also won't because the fix / workaround was a change in behaviour, so it's a breaking change

**Comment by bpeck81** [comment]
I believe I'm seeing this issue in react native when the app first launches. I'm positive I prefetch data and the prefetching resolves. When I query that data with the same key after navigating to a new page with the stack navigator, it returns undefined for a while. If I launch the app and wait about 5 seconds before navigating to the screen, the data is there. So for me it is shortly after app launch, which unfortunately is the most critical time for my app to work quickly.

I'm using v4. I haven't shifted to v5 because it requires a lot of rewriting onSuccess logic. Does v5 resolve the issue?

**Comment by TkDodo** [comment]
@lundstromdavid it seems like you're clearing the cache while the query is in-flight, I don't know why you would do something like that. `clear()` is not the same as `resetQueries()` because clear does not inform observers about the change

**Comment by lundstromdavid** [comment]
@TkDodo Thanks for the reply! Should have read up more before. Although it does seem quite unintuitive for me that `clear()` can stall in-flight queries in a loading state, but what I do know :man_shrugging:

**Comment by shuweili1122** [comment]
can confirm this indeed happens with `react: 18.2`, and `@tanstack/react-query 4.36.1` on `post` request when running `jest` test with `axios 1.6.2` and `axios-mock-adapter 1.22.0`, still happens after upgrading `@tanstack/react-query` to `5.17.19`, I can see the result in useQuery, but it never returns to the component, and the `status` is always `loading`; hence used vanilla `axios` for my failed test and everything passes.

Also found a very hacky workaround if you don't want to replace `@tanstack/react-query` with other libraries: basically just do a data fetching with `axios-mock-adapter` in `beforeAll` or `beforeEach`, then run your regular test, and in your test block, do a data fetching with axios-mock-adapter again... your test will then have about 1/2 chance of passing... super confusing bug

**Comment by TkDodo** [comment]
One thing you can try on v5 is to disable our batching and see if that helps:

```
import { notifyManager } from '@tanstack/react-query'

// invoke callback instantly
notifyManager.setScheduler((cb) => c())
```

see: https://tanstack.com/query/latest/docs/reference/notifyManager#notifymanagersetscheduler

**Comment by Codyooo** [comment]
I would like to recommend changing the default setting of the networkMode configuration in React Query to 'always'. Given the library's widespread use, with over 200 million downloads per week, the current default setting may lead to significant issues.

Recently, we encountered a bug where the system did not make any network requests due to the networkMode being set to 'online'. This issue arose because the browser's navigator.onLine API inaccurately reported the network status as offline, even though the connection was active. The impact of this bug is particularly severe as it can cause key pages to remain stuck on a loading screen without any indication of the problem. Considering that React Query is a critical part of many projects, such errors can have drastic consequences, including potential job losses, especially if key stakeholders encounter these issues firsthand.

It can be challenging to identify this issue without delving into the source code, making it a silent but deadly problem. To enhance reliability and prevent such problematic scenarios, I strongly urge you to consider setting the default networkMode to 'always'. This change would ensure that network requests are made regardless of the navigator.onLine status, thus providing a more consistent and dependable user experience.

Thank you for considering this suggestion. I believe it will significantly improve the robustness of React Query and safeguard many developers and businesses from facing similar issues.

**Comment by TkDodo** [comment]
you can set the scheduler inside cypress tests only.

**Comment by code-with-shahzad** [comment]
Has any face this issue, My queries stuck infinite loading after, I have to restart to app to get data. I am using React Query V: 

    "@tanstack/react-query": "5.59.19",

**Comment by dylan-chong** [comment]
I'm still using super old react-query 3.39.3 and encountered the stuck in loading issue
We found that we were firing 2 identical requests from 2 different components but the query keys were slightly different, when they should be the same to dedupe the request
It seemed that the bug in react-query would be something like the two identical requests can clobber the same key in a map, preventing one of the components receiving the update
Making the query keys the same fixed it
